### ISSUE TITLE
Search with OCR / image recognition

### ISSUE THREAD

**Issue by hemanthsoni** [SergiX44/XBackBone#21]
It would be awesome to integrate a search function with something like [Tesseract OCR](https://github.com/tesseract-ocr/tesseract) or Google's [OCR](https://cloud.google.com/vision/docs/ocr) and [Vision](https://cloud.google.com/vision/) APIs.

This would let users search their screenshots, etc. based on the text on the screen at the time of the screenshot, or content of the image upload (e.g., "zebra", "dog", etc.).

**Comment by codemicro** [comment]
Google's Vision and OCR APIs require (similarly to Google Cloud) a Google Cloud account and project, however the project appears to require billing, meaning you cannot use any free credit on it and hence need to pay. The advantage of Google Cloud is that you can have everything behind an API, meaning it makes it a whole lot simpler to bundle within XBackBone provided the user supplies credentials. Google also provides their own PHP wrapper extension.

Tesseract however requires installation onto the host machine. There exists a wrapper library for it ([thiagoalessio/tesseract-ocr-for-php](https://github.com/thiagoalessio/tesseract-ocr-for-php)) but this still depends on having Tesseract installed. You don't have to pay for Tesseract, but it will create extra complexeties during installation and setup.

Either way this would have to be an optional feature. In this scenario I'd lean towards Tesseract because it's open source and free. I'd be interested in helping out with implementing this but it's going to need more research first and hopefully for @SergiX44 to make some decisions made about how it'll be implemented.

**Comment by SergiX44** [comment]
I agree on the tesseract way, because the google vision api is neither open source nor free, and is not self hosted.
Before this, a tag system is needed to assign multiple keywords to an upload. After that, a batch system that calls tesseract, extract the text content from the images, use something like rake to extract only the keywords and then save all to the database.

**Comment by codemicro** [comment]
I've done some testing!

https://gist.github.com/codemicro/e88d6deb21dd1792ed230e6510bddb5a

I used [Donatello-za/rake-php-plus](https://github.com/Donatello-za/rake-php-plus) for the keyword recognition and [thiagoalessio/tesseract-ocr-for-php](https://github.com/thiagoalessio/tesseract-ocr-for-php) for the Tesseract integration. An issue with this is that because Tesseract requires an image file as an command line argument, this does too, meaning I've got to save the file to a temporary file before running the OCR command - perhaps just something to bear in mind.

The images were chosen for a few reasons:
* One and two were chosen because they're just bodies of text - it did well with these
* Three was chosen to test how Tesseract would work with screenshots that aren't just text, and the answer is not too good.
* Four was chosen to test an image with no text
* Five was chosen for something with a small amount of text

There's potential, but there's too much potential for Tesseract to spit out rubbish at this point. Perhaps another OCR engine is needed?

**Comment by SergiX44** [comment]
Thank you for the testing!

I see also other problems: rake is not language agnostic, and only support 4-5 langs, and also, tesseract need to know the target language in the image to recognize. Maybe tuning some options like psm and set the target language would reduce the amount of potential rubbish.

**Comment by codemicro** [comment]
The text in these images is all in English, which is the default language for both Tesseract and RAKE. I agree that the lack of languages is potentially problematic (and maybe support would only be offered for those languages) and a setting would be needed similarly to that of the overall language selection in XBackBone.

Following some of the instructions in [this wiki page](https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality) I was able to significantly improve the output generated by the third image by inverting the colours.

Image: https://www.tdpain.net/sharex/suxE1/luHUlevoHi372.png/raw

Output:

      @everyone WE NEED YOU

      If you're free, please LOGON NOW onto the live network

      7S) 3 ™ 20 (MIO) MO) 7 oF 7 89 07 BENS OS M7 Zt 84 #2 Bi
      a2ia

      If server says "max clients reached’, please try another one
      mo 3 @3°44

      We are slowly increasing server limits to distribute load evenly across all
      ™14 #4683 13 M3 8e@

      PILOTS PREFERRED
      84 d9 @6 M5 %7 45 06 #486

      For any issues, please use the respective forum sections. If it's a pilot client issue, report it in the
      for your pilot client. Ifit's an AFV the AFV Section. Many thanks for your patience and help all throughtout

      Sc@7a

PSM didn't do a whole lot to help either - I saw no noticiable change.

**Comment by gerroon** [comment]
I think having some form of ocr on the images is better than no ocr, at least we can get some automated tagging at think even a %50 success would be good with tagging.

**Comment by enescaglayan** [comment]
> The text in these images is all in English, which is the default language for both Tesseract and RAKE. I agree that the lack of languages is potentially problematic (and maybe support would only be offered for those languages) and a setting would be needed similarly to that of the overall language selection in XBackBone.
> 
> Following some of the instructions in [this wiki page](https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality) I was able to significantly improve the output generated by the third image by inverting the colours. Finding a way to automatically decide if the image should be inverted before OCR might be beneficial.
> 
> Image: https://www.tdpain.net/sharex/suxE1/luHUlevoHi372.png/raw
> 
> Output:
> 
> ```
>   @everyone WE NEED YOU
> 
>   If you're free, please LOGON NOW onto the live network
> 
>   7S) 3 ™ 20 (MIO) MO) 7 oF 7 89 07 BENS OS M7 Zt 84 #2 Bi
>   a2ia
> 
>   If server says "max clients reached’, please try another one
>   mo 3 @3°44
> 
>   We are slowly increasing server limits to distribute load evenly across all
>   ™14 #4683 13 M3 8e@
> 
>   PILOTS PREFERRED
>   84 d9 @6 M5 %7 45 06 #486
> 
>   For any issues, please use the respective forum sections. If it's a pilot client issue, report it in the
>   for your pilot client. Ifit's an AFV the AFV Section. Many thanks for your patience and help all throughtout
> 
>   Sc@7a
> ```
> 
> PSM didn't do a whole lot to help either - I saw no noticiable change.

Whats the discord server you got screenshot from?

**Comment by codemicro** [comment]
I don't see how that's relevant.

**Comment by enescaglayan** [comment]
Yes

**Comment by FlintSH** [comment]
Personally, I switched to XBackBone & ShareX coming from Gyazo, and one of Gyazo's best features was automatic OCR and search capabilities. This would be really awesome!
